content/articles/api-authentication-with-django-knox-and-postman-testing/index.md

```diff
@@ -0,0 +1,343 @@
+
+from .api import LeadViewset
+
+router = routers.DefaultRouter()
+router.register('leads', LeadViewset, 'leads')
+
+urlpatterns = router.urls
+```
+Whenever there are `HTTP GET` calls with `/leads` endpoint, the API fetches the client all serialized leads.
+
+Let us register the endpoint with the project by going into the `urls.py` of the `Leads`folder.
+
+```python
+from django.contrib import admin
+from django.urls import path, include
+
+urlpatterns = [
+    path('api/', include('leadsapi.urls')),
+    path('', include('accounts.urls'))
+]
+```
+Note that the `account.urls` shall be used later in the tutorial.
+
+### Securing the endpoints with Django-Knox
+
+Now that we have built the rest API, it is time to secure it by allowing only the authenticated users to perform HTTP `GET`, `POST` methods.
+
+We will make a fresh app called `account` which I believe we have demonstrated above. Follow the procedures, make the new app and add it to the settings.
+
+As we have said that Knox provides a token to every logged-in user, for them to access the secured data. 
+
+We will be serializing the data of the user by making use of the Django built-in `User` model. This will be done inside the `serializers.py` in the new app created.
+
+```python
+from django.contrib.auth import authenticate
+from django.contrib.auth.models import User
+from django.db import models
+from django.db.models import fields
+from rest_framework import serializers
+
+class UserSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = User
+        fields = ('id', 'username', 'email')
+
+
+class RegisterSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = User
+        fields = ('id', 'username', 'email', 'password')
+        extra_kwargs = {'password': {'write_only': True}}
+
+    def create(self, validated_data):
+        user = User.objects.create_user(validated_data['username'], validated_data['email'], validated_data['password'])
+        return user
+
+
+class LoginSerializer(serializers.Serializer):
+    username = serializers.CharField()
+    password = serializers.CharField()
+
+    def validate(self, data):
+        user = authenticate(**data)
+        if user and user.is_active:
+            return user
+        raise serializers.ValidationError('Incorrect Credentials Passed.')
+```
+
+The `UserSerializer` class maps the `User` model and extracts the mentioned fields for the serialization.
+
+The `RegisterSerialize` class does the same thing but added the password field with the modality of ``write_only`. 
+
+While the `LoginSerializer`class does the trick of the authentication. The method called `validate` requires the data as the argument which passes it to the `authenticate` method from the rest framework. 
+
+By doing so, the user will be verified with their token that is already inside the data passed. If the token provided matches, the user will be logged in. But if otherwise,  a validation error will be sent.
```

Comment:
Change to " We will verify the user with their token that is already inside the data passed. If the token provided matches, the user would be logged in. But if otherwise, the application will send a validation error."